Use NextPage type from Next.js

singingwolfboy opened the pull request:
We can use the `NextPage` type from Next.js to tighten up our types around pages. Notably, this also cascades down to the `.getInitialProps` functions, which weren't defined properly according to Next.js's types, so I tried to correct them.

singingwolfboy commented on @app/client/src/pages/login.tsx:
```diff
@@ -101,14 +101,17 @@ export default function Login({ next: rawNext }: LoginProps) {
 
-Login.getInitialProps = ({ query }: NextPageContext) => {
+Login.getInitialProps = async ({ query }) => {
+  const next = typeof query.next === "string" ? query.next : query.next[0];
```
According to the type definitions from Next.js, the values on the `query` object are of type `string | string[]`. In retrospect, this makes sense: you can repeat query parameters in a URL, like this: `http://www.example.com/login?next=foo&next=bar`. I suppose in that case, `query.next` would equal `["foo", "bar"]`.

One thing I _don't_ understand is why the type isn't `string | string[] | undefined`. Either Next.js returns the empty string to represent a query param that isn't in the URL, or the type definition is missing the `undefined` type. Do you know the situation here?

benjie replied:
Assuming it's a map `{[key: string]: string | string[]}` then it's not possible that every string key could be defined, so `undefined` is implicit I think? :man_shrugging:

singingwolfboy commented:
I added a `firstIfArray` utility function, and refactored the `getInitialProps()` functions to use it where appropriate. That should handle `undefined` references as well.

I also refactored the `ErrorPage` component to use the `NextPage` type as well -- I missed it the first time around. That required some larger code changes to make it work nicely with TypeScript types. Please let me know if you see any issues!
